src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const allStates = [];
+  const stateCopy = { ...state };
+  const ADD_PROPERTIES_TYPE = 'addProperties';
+  const REMOVE_PROPERTIES_TYPE = 'removeProperties';
+  const CLEAR_PROPERTIES_TYPE = 'clear';
+
+  for (const action of actions) {
+    switch (action.type) {
+      case ADD_PROPERTIES_TYPE:
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case REMOVE_PROPERTIES_TYPE:
+        action.keysToRemove.forEach(key => {
+          delete stateCopy[key];
+        });
+        break;
+
+      case CLEAR_PROPERTIES_TYPE:
+        Object.keys(stateCopy).forEach(key => {
+          delete stateCopy[key];
+        });
```

Comment:
as a suggestion you can assign an empty object to clear the state
```suggestion
        stateCopy = {};
```